modules/dnn/src/cuda4dnn/cxx_utils/is_iterator.hpp

```diff
@@ -0,0 +1,31 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_DNN_CUDA4DNN_CXX_UTILS_IS_ITERATOR_HPP
+#define OPENCV_DNN_CUDA4DNN_CXX_UTILS_IS_ITERATOR_HPP
+
+#include <iterator>
+#include <type_traits>
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace cxx_utils {
+
+    namespace detail {
+        template <class T, class Tag, class = void>
+        struct is_iterator_helper : std::false_type {};
+
+        template <class T, class Tag>
+        struct is_iterator_helper<T, Tag,
+                typename std::enable_if<std::is_base_of<Tag, typename std::iterator_traits<T>::iterator_category>::value, void>::type
```

Comment:
I think this is undefined behaviour. The C++11 standard does not specify what happens when `std::iterator_traits` is instantiated with a non-iterator type. It also does not provide any guarantee that the `std::iterator_traits` SFINAE friendly. Hence, the identified line of code is non-conforming code. It's conforming-code in C++17 however.

Reply:
The standard doesn't say that it's implementation-defined. It only mentions what should happen when `std::iterator_traits` is instantiated for an iterator. So it's UB.

Even though not explicitly stated in the standard, an implementation could decide what should happen. It turns out gcc, clang and msvc are in our favor.

Reply:
Normally, trying to instantiate a template that doesn't have any applicable specialization is definitely not UB.  Unless the standard explicitly calls this out as UB for `std::iterator_traits` (which seems nutty to me, but I haven't looked to see what the standard says) it ought to be just like any other case of using a template that doesn't have any appropriate specialization: you get a compile time error.